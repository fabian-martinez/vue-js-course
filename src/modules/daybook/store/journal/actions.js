export const loadEntries = async ({commit}) => {

}
export const updateEntry = async ({commit}) => {

}
export const createEntru = async ({commit}) => {

}