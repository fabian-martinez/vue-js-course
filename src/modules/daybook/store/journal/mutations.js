export const setEntries = (state) => {

}
export const updateEntriy = (state) => {

}
export const addEntry  = (state) => {

}