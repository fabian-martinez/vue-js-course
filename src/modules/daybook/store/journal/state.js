export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text:'Nulla occaecat laboris mollit irure non fugiat minim cupidatat irure aliquip do. Commodo minim ex tempor ex do cillum ut. Excepteur amet quis magna reprehenderit enim aliqua commodo quis reprehenderit do voluptate dolore qui minim. Id minim dolor magna minim. Consequat consequat consequat fugiat laboris tempor qui aute duis culpa duis occaecat elit.',
            picture: null
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text:'Nulla occaecat laboris  irure non fugiat minim cupidatat irure aliquip do. Commodo minim ex tempor ex do cillum ut. Excepteur amet quis magna reprehenderit enim aliqua commodo quis reprehenderit do voluptate dolore qui minim. Id minim dolor magna minim. Consequat consequat consequat fugiat laboris tempor qui aute duis culpa duis occaecat elit.',
            picture: null
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text:'Nulla occaecat laboris mollit irure non fugiat minim cupidatat irure aliquip do. Commodo minim ex tempor ex do cillum ut. Excepteur amet quis magna reprehenderit enim aliqua commodo quis reprehenderit do voluptate dolore qui minim. Id minim dolor magna minim. Consequat consequat consequat fugiat laboris tempor qui aute duis culpa duis occaecat elit.',
            picture: null
        },
    ]
})