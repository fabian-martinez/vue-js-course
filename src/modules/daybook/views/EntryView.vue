<template lang="">
    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{ day }}</span>
                <span class="mx-1 fs-3">{{ month }}</span>
                <span class="mx-2 fs-4 fw-ight">{{ year }}, {{ weekDay }}</span>
            </div>
            <div>
                <button v-if="entry.id" class="btn btn-danger mx-2" @click="onDeleteEntry">
                    Borrar
                    <i class="fa fa-trash-alt"></i>
                </button>
                <input type="file" 
                    @change="onSelectedImage" 
                    ref="imageSelector"
                    v-show="false"
                    accept="image/png image/jgp"
                    >
                <button class="btn btn-primary" @click="onSelectImage">
                    Subir fotografia
                    <i class="fa fa-upload"></i>
                </button>
            </div>
        </div>
        <hr>
        <div class="d-flex flex-column px-3 h-75">
             <textarea placeholder="Que sucedio hoy?" v-model="entry.text"></textarea>
        </div>
        <fab icon="fa-save" @on:click="saveEntry"></fab>
        <img v-if="entry.picture && !localImage"
        :src="entry.picture" 
        alt="entry-picture"
        class="img-thumbnail"
        >
        <img v-if="localImage"
        :src="localImage" 
        alt=''
        class="img-thumbnail"
        >
    </template>
</template>
<script>
import { defineAsyncComponent } from '@vue/runtime-core';
import { mapActions, mapGetters } from 'vuex'
import Swal from 'sweetalert2'

import getDayMonthYear from '@/modules/daybook/helpers/getDayMonthYear'
import uploadImages from '@/modules/daybook/helpers/uploadImages'


export default {
    name: 'EntryView',
    props:{
        id:{
            type: String,
            require: true,
        } 
    },
    computed: {
        ...mapGetters( 'journal', ['getEntryById']  ),
        weekDay(){
            const { weekDay } = getDayMonthYear( this.entry.date )
            return weekDay
        },
        day(){
            const { day } = getDayMonthYear( this.entry.date )
            return day
        },
        month(){
            const { month } = getDayMonthYear( this.entry.date )
            return month
        },
        year(){
            const { year } = getDayMonthYear( this.entry.date )
            return year
        }
    },
    components:{
        Fab: defineAsyncComponent(() => import(/* webpackChunkName: "daybook" */ '../components/Fab.vue'))
    },
    data() {
        return {
            entry: null,
            localImage: null,
            file: null,
        }
    },
    methods: {
        ...mapActions( 'journal', ['updateEntry', 'createEntry', 'deleteEntry'] ),
        loadEntry() {
            let entry
            if (this.id === 'new'){
                entry = {
                    date: new Date().toString(),
                    text: ''
                }
            }else{
                entry = this.getEntryById(this.id)
                if (!entry) return this.$router.push({ name: 'no-entry' })
            }
            this.entry = entry
            this.localImage = null
            this.file = null
        },
        async saveEntry() {
            new Swal({
                title: 'Espere porfavor',
                allowOutsideClick:true
            })
            Swal.showLoading()
            this.entry.picture = await uploadImages(this.file)
            if ( this.entry.id ) {
                this.updateEntry(this.entry)
            } else {
                const id = await this.createEntry(this.entry)
                this.$router.push({ name: 'entry', params:{ id }})
            }
            Swal.fire('Guardado','Entrada registrada exitosamente','success')
        },
        async onDeleteEntry() {
            const { isConfirmed } = await Swal.fire({
                title:'Â¿Esta seguro de borrar esta entrada?',
                text: 'Una vez borrada no se puede recuperar',
                showDenyButton:true,
                confirmButtonText:'Si, estoy seguro'
            })
            if (!isConfirmed) return 
            Swal.fire({
                title: 'Espere porfavor',
                allowOutsideClick:true
            })
            Swal.showLoading()

            const id = await this.deleteEntry(this.entry.id)
            console.log(`La entrada ${id} fue eliminada`);
            this.$router.push({ name:'no-entry'})
            this.file = null
            Swal.fire('Eliminacion exitos','Entrada registrada exitosamente','success')  
        },
        onSelectedImage($event){
            const file = $event.target.files[0]
            if(!file){
                this.localImage = null
                this.file = null
                return
            }
            this.file = file
            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL(file)  
        },
        onSelectImage(){
            this.$refs.imageSelector.click();
        }
    },
    created() {
        this.loadEntry()
    },
    watch: {
        id( value, oldValue ){
            this.loadEntry()
        }
    }
}
</script>
<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;

    &:focus{
        outline: none;
    }
}
img {
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.3);
}
</style>