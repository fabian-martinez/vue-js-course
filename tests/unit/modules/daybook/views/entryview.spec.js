import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";

import Swal from 'sweetalert2'

import { journalState } from "../../../mock-data/test-journal-state";
import journal from "@/modules/daybook/store/journal";

import EntryView from "@/modules/daybook/views/EntryView.vue";
 

jest.mock('sweetalert2', () => ({
    showLoading: jest.fn(),
    fire: jest.fn(), 
})) 

describe('Pruebas del EntrieView', () => {
    
    const mockRouter = { 
        push: jest.fn() 
    }

    const createVuexStore = (initialState) => 
        createStore({
            modules: {
                journal: {
                    ...journal,
                    state: { ...initialState }
                }
            }
        })

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn() 

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks() 
        wrapper = shallowMount(EntryView, {
            props:{
                id:'-NGDwnVJ0duTetAhtLMB'
            },
            global:{
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })  
    })  

    it('Debe sacar al usuario si el id no existe', () => {
        const wrapper = shallowMount(EntryView, {
            props:{
                id: 'ID-NO-EXISTE'
            },
            global:{
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
        
        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })
    });

    it('Debe mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled() 
    });

    it('debe borrar la entrada y salir', async() => {
        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }) )
        await wrapper.find('.btn-danger').trigger('click')

        expect( Swal.fire ).toHaveBeenCalledWith({
            title:'Â¿Esta seguro de borrar esta entrada?',
            text: 'Una vez borrada no se puede recuperar',
            showDenyButton:true,
            confirmButtonText:'Si, estoy seguro'
        })

        expect( store.dispatch ).toHaveBeenCalledWith("journal/deleteEntry", "-NGDwnVJ0duTetAhtLMB")
        expect( mockRouter.push ).toHaveBeenCalled()

    });
});