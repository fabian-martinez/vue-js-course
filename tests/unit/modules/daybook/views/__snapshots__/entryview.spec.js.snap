// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del EntrieView Debe mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">6</span><span class="mx-1 fs-3">noviembre</span><span class="mx-2 fs-4 fw-ight">2022, domingo</span></div>
  <div><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><input type="file" accept="image/png image/jgp" style="display: none;"><button class="btn btn-primary"> Subir fotografia <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Que sucedio hoy?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub>
<!--v-if-->
<!--v-if-->
`;
